File: collect_app/src/androidTest/java/org/odk/collect/android/formfilling/FieldListUpdateTest.java
package org.odk.collect.android.formfilling;

import android.Manifest;
import android.content.Intent;
import android.content.res.AssetManager;
import android.os.Environment;
import androidx.test.InstrumentationRegistry;
import androidx.test.core.app.ApplicationProvider;
import androidx.test.espresso.intent.rule.IntentsTestRule;
import androidx.test.rule.GrantPermissionRule;
import androidx.test.runner.AndroidJUnit4;
import android.view.View;

import org.apache.commons.io.IOUtils;

import org.hamcrest.Description;
import org.hamcrest.Matcher;
import org.hamcrest.TypeSafeMatcher;
import org.junit.BeforeClass;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.odk.collect.android.R;
import org.odk.collect.android.activities.FormEntryActivity;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.UUID;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.action.ViewActions.longClick;
import static androidx.test.espresso.action.ViewActions.replaceText;
import static androidx.test.espresso.assertion.PositionAssertions.isCompletelyAbove;
import static androidx.test.espresso.assertion.PositionAssertions.isCompletelyBelow;
import static androidx.test.espresso.assertion.ViewAssertions.doesNotExist;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withClassName;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;

import static org.hamcrest.Matchers.allOf;
import static org.hamcrest.Matchers.endsWith;
import static org.hamcrest.core.StringStartsWith.startsWith;
import static org.odk.collect.android.activities.FormEntryActivity.EXTRA_TESTING_PATH;

@RunWith(AndroidJUnit4.class)
/**
 * TODO: consolidate boilerplate with {@link org.odk.collect.android.AllWidgetsFormTest}? Challenge
 * is that much of it is static and needs to have the form path.
 */
public class FieldListUpdateTest {
    private static final String FORMS_PATH = Environment.getExternalStorageDirectory().getPath() + "/odk/forms/";
    private static final String FIELD_LIST_TEST_FORM = "fieldlist-updates.xml";

    @Rule
    public FormEntryActivityTestRule activityTestRule = new FormEntryActivityTestRule();

    @Rule
    public GrantPermissionRule permissionRule = GrantPermissionRule.grant(Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.WRITE_EXTERNAL_STORAGE);

    @BeforeClass
    public static void copyFormToSdCard() throws IOException {
        String pathname =  FORMS_PATH + FIELD_LIST_TEST_FORM;

        AssetManager assetManager = InstrumentationRegistry.getContext().getAssets();
        InputStream inputStream = assetManager.open(FIELD_LIST_TEST_FORM);

        File outFile = new File(pathname);
        OutputStream outputStream = new FileOutputStream(outFile);

        IOUtils.copy(inputStream, outputStream);
    }

    @Test
    public void relevanceChangeAtEnd_ShouldToggleLastWidgetVisibility() {
        onView(withId(R.id.menu_goto)).perform(click());
        onView(withId(R.id.menu_go_up)).perform(click());
        // TODO: what is with those hidden views with the same text?
        onView(allOf(withText("Single relevance at end"), isDisplayed())).perform(click());
        onView(withText("Source1")).perform(click());

        onView(withIndex(withClassName(endsWith("EditText")), 0)).perform(replaceText(""));
        onView(withText("Target1")).check(doesNotExist());
        onView(withIndex(withClassName(endsWith("EditText")), 0)).perform(replaceText("A"));
        onView(withText("Target1")).check(matches(isDisplayed()));
        onView(withText("Target1")).check(isCompletelyBelow(withText("Source1")));

        onView(withIndex(withClassName(endsWith("EditText")), 0)).perform(replaceText(""));
        onView(withText("Target1")).check(doesNotExist());
    }

    @Test
    public void relevanceChangeAtBeginning_ShouldToggleFirstWidgetVisibility() {
        onView(withId(R.id.menu_goto)).perform(click());
        onView(withId(R.id.menu_go_up)).perform(click());
        onView(allOf(withText("Single relevance at beginning"), isDisplayed())).perform(click());
        onView(withText("Source2")).perform(click());

        onView(withIndex(withClassName(endsWith("EditText")), 0)).perform(replaceText(""));
        onView(withText("Target2")).check(doesNotExist());
        onView(withIndex(withClassName(endsWith("EditText")), 0)).perform(replaceText("A"));
        onView(withText("Target2")).check(matches(isDisplayed()));
        onView(withText("Target2")).check(isCompletelyAbove(withText("Source2")));

        onView(withIndex(withClassName(endsWith("EditText")), 1)).perform(replaceText(""));
        onView(withText("Target2")).check(doesNotExist());
    }

    @Test
    public void relevanceChangeInMiddle_ShouldToggleMiddleWidgetVisibility() {
        onView(withId(R.id.menu_goto)).perform(click());
        onView(withId(R.id.menu_go_up)).perform(click());
        onView(allOf(withText("Single relevance in middle"), isDisplayed())).perform(click());
        onView(withText("Source3")).perform(click());

        onView(withIndex(withClassName(endsWith("EditText")), 0)).perform(replaceText(""));
        onView(withText("Target3")).check(doesNotExist());
        onView(withIndex(withClassName(endsWith("EditText")), 0)).perform(replaceText("A"));
        onView(withText("Target3")).check(matches(isDisplayed()));
        onView(withText("Target3")).check(isCompletelyBelow(withText("Source3")));
        onView(withText("Target3")).check(isCompletelyAbove(withText("Filler3")));

        onView(withIndex(withClassName(endsWith("EditText")), 0)).perform(replaceText(""));
        onView(withText("Target3")).check(doesNotExist());
    }

    @Test
    public void longPress_ShouldClearAndUpdate() {
        onView(withId(R.id.menu_goto)).perform(click());
        onView(withId(R.id.menu_go_up)).perform(click());
        onView(allOf(withText("Single relevance in middle"), isDisplayed())).perform(click());
        onView(withText("Source3")).perform(click());

        onView(withIndex(withClassName(endsWith("EditText")), 0)).perform(replaceText(""));
        onView(withText("Target3")).check(doesNotExist());
        onView(withIndex(withClassName(endsWith("EditText")), 0)).perform(replaceText("A"));
        onView(withText("Target3")).check(matches(isDisplayed()));

        onView(withText("Source3")).perform(longClick());
        onView(withText(R.string.clear_answer)).perform(click());
        onView(withText(R.string.discard_answer)).perform(click());
        onView(withIndex(withClassName(endsWith("EditText")), 0)).check(matches(withText("")));
        onView(withText("Target3")).check(doesNotExist());
    }

    @Test
    public void changeInValueUsedInLabel_ShouldChangeLabelText() {
        onView(withId(R.id.menu_goto)).perform(click());
        onView(withId(R.id.menu_go_up)).perform(click());
        onView(allOf(withText("Label change"), isDisplayed())).perform(click());
        onView(withText(startsWith("Hello"))).perform(click());

        String name = UUID.randomUUID().toString();

        onView(withIndex(withClassName(endsWith("EditText")), 1)).perform(replaceText(""));
        onView(withText("Hello, , how are you today?")).check(matches(isDisplayed()));
        onView(withIndex(withClassName(endsWith("EditText")), 1)).perform(replaceText(name));
        onView(withText("Hello, " + name + ", how are you today?")).check(matches(isDisplayed()));
        onView(withIndex(withClassName(endsWith("EditText")), 1)).perform(replaceText(""));
        onView(withText("Hello, , how are you today?")).check(matches(isDisplayed()));
    }

    @Test
    public void changeInValueUsedInHint_ShouldChangeHintText() {
        onView(withId(R.id.menu_goto)).perform(click());
        onView(withId(R.id.menu_go_up)).perform(click());
        onView(allOf(withText("Hint change"), isDisplayed())).perform(click());
        onView(withText(startsWith("What is your"))).perform(click());

        String name = UUID.randomUUID().toString();

        onView(withIndex(withClassName(endsWith("EditText")), 0)).perform(replaceText(""));
        onView(withText("Please don't use your calculator, !")).check(matches(isDisplayed()));
        onView(withIndex(withClassName(endsWith("EditText")), 0)).perform(replaceText(name));
        onView(withText("Please don't use your calculator, " + name + "!")).check(matches(isDisplayed()));
        onView(withIndex(withClassName(endsWith("EditText")), 0)).perform(replaceText(""));
        onView(withText("Please don't use your calculator, !")).check(matches(isDisplayed()));
    }

    /** TODO: calculation doesn't seem to be updated whether or not there's a fieldlist.
    @Test
    public void changeInValueUsedInOtherField_ShouldChangeValue() {
        onView(withId(R.id.menu_goto)).perform(click());
        onView(withId(R.id.menu_go_up)).perform(click());
        onView(allOf(withText("Value change"), isDisplayed())).perform(click());
        onView(withText(startsWith("What is your"))).perform(click());

        String name = UUID.randomUUID().toString();

        onView(withIndex(withClassName(endsWith("EditText")), 0)).perform(replaceText(""));
        onView(withIndex(withClassName(endsWith("EditText")), 1)).check(matches(withText("0")));
        onView(withIndex(withClassName(endsWith("EditText")), 0)).perform(replaceText(name));
        onView(withIndex(withClassName(endsWith("EditText")), 1)).check(matches(withText(name.length())));
        onView(withIndex(withClassName(endsWith("EditText")), 0)).perform(replaceText(""));
        onView(withIndex(withClassName(endsWith("EditText")), 1)).check(matches(withText("0")));
    }
    **/

    @Test
    public void selectionChangeAtFirstCascadeLevel_ShouldUpdateNextLevels() {
        onView(withId(R.id.menu_goto)).perform(click());
        onView(withId(R.id.menu_go_up)).perform(click());
        onView(allOf(withText("Cascading select"), isDisplayed())).perform(click());
        onView(withText(startsWith("Level1"))).perform(click());

        // No choices should be shown for levels 2 and 3 when no selection is made for level 1
        onView(withText("A1")).check(doesNotExist());
        onView(withText("B1")).check(doesNotExist());
        onView(withText("C1")).check(doesNotExist());
        onView(withText("A1A")).check(doesNotExist());

        // Selecting C for level 1 should only reveal options for C at level 2
        onView(withText("C")).perform(click());
        onView(withText("A1")).check(doesNotExist());
        onView(withText("B1")).check(doesNotExist());
        onView(withText("C1")).check(matches(isDisplayed()));
        onView(withText("A1A")).check(doesNotExist());

        // Selecting A for level 1 should reveal options for A at level 2
        onView(withText("A")).perform(click());
        onView(withText("A1")).check(matches(isDisplayed()));
        onView(withText("A1A")).check(doesNotExist());
        onView(withText("B1")).check(doesNotExist());
        onView(withText("C1")).check(doesNotExist());

        // Selecting A1 for level 2 should reveal options for A1 at level 3
        onView(withText("A1")).perform(click());
        onView(withText("A1A")).check(matches(isDisplayed()));
        onView(withText("B1")).check(doesNotExist());
        onView(withText("C1")).check(doesNotExist());
    }

    // TODO: figure out why the third level isn't cleared. Wondering whether there might be an issue
    // with populateDynamicChoices
    //    @Test
    //    public void clearingParentSelect_ShouldUpdateAllDependentLevels() {
    //        onView(withId(R.id.menu_goto)).perform(click());
    //        onView(withId(R.id.menu_go_up)).perform(click());
    //        onView(allOf(withText("Cascading select"), isDisplayed())).perform(click());
    //        onView(withText(startsWith("Level1"))).perform(click());
    //
    //        onView(withText("A")).perform(click());
    //        onView(withText("A1")).perform(click());
    //        onView(withText("A1B")).perform(click());
    //
    //        onView(withText("A")).perform(longClick());
    //        onView(withText(R.string.clear_answer)).perform(click());
    //        onView(withText(R.string.discard_answer)).perform(click());
    //
    //        onView(withIndex(withClassName(endsWith("RadioButton")), 0)).check(matches(isNotChecked()));
    //        onView(withText("A1")).check(doesNotExist());
    //        onView(withText("A1B")).check(doesNotExist());
    //    }

    @Test
    public void selectionChangeAtOneCascadeLevelWithMinimalAppearance_ShouldUpdateNextLevels() {
        onView(withId(R.id.menu_goto)).perform(click());
        onView(withId(R.id.menu_go_up)).perform(click());
        onView(allOf(withText("Cascading select minimal"), isDisplayed())).perform(click());
        onView(withText(startsWith("Level1"))).perform(click());

        // No choices should be shown for levels 2 and 3 when no selection is made for level 1
        onView(withText("A1")).check(doesNotExist());
        onView(withText("B1")).check(doesNotExist());
        onView(withText("C1")).check(doesNotExist());
        onView(withText("A1A")).check(doesNotExist());

        // Selecting C for level 1 should only reveal options for C at level 2
        onView(withIndex(withText(R.string.select_one), 0)).perform(click());
        onView(withText("C")).perform(click());
        onView(withText("A1")).check(doesNotExist());
        onView(withText("B1")).check(doesNotExist());
        onView(withIndex(withText(R.string.select_one), 0)).perform(click());
        onView(withText("C1")).perform(click());
        onView(withText("A1A")).check(doesNotExist());

        // Selecting A for level 1 should reveal options for A at level 2
        onView(withText("C")).perform(click());
        onView(withText("A")).perform(click());
        onView(withIndex(withText(R.string.select_one), 0)).perform(click());
        onView(withText("A1")).check(matches(isDisplayed()));
        onView(withText("A1A")).check(doesNotExist());
        onView(withText("B1")).check(doesNotExist());
        onView(withText("C1")).check(doesNotExist());

        // Selecting A1 for level 2 should reveal options for A1 at level 3
        onView(withText("A1")).perform(click());
        onView(withIndex(withText(R.string.select_one), 0)).perform(click());
        onView(withText("A1A")).check(matches(isDisplayed()));
        onView(withText("B1A")).check(doesNotExist());
        onView(withText("B1")).check(doesNotExist());
        onView(withText("C1")).check(doesNotExist());
    }

    private static class FormEntryActivityTestRule extends IntentsTestRule<FormEntryActivity> {
        FormEntryActivityTestRule() {
            super(FormEntryActivity.class);
        }

        @Override
        protected Intent getActivityIntent() {
            Intent intent = new Intent(ApplicationProvider.getApplicationContext(), FormEntryActivity.class);
            intent.putExtra(EXTRA_TESTING_PATH, FORMS_PATH + FIELD_LIST_TEST_FORM);

            return intent;
        }
    }

    // https://stackoverflow.com/a/39756832
    public static Matcher<View> withIndex(final Matcher<View> matcher, final int index) {
        return new TypeSafeMatcher<View>() {
            int currentIndex;

            @Override
            public void describeTo(Description description) {
                description.appendText("with index: ");
                description.appendValue(index);
                matcher.describeTo(description);
            }

            @Override
            public boolean matchesSafely(View view) {
                return matcher.matches(view) && currentIndex++ == index;
            }
        };
    }
}

File: collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/FieldListUpdateTest.kt
package org.odk.collect.android.feature.formentry

import android.app.Activity
import android.app.Instrumentation
import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import androidx.test.core.app.ApplicationProvider
import androidx.test.espresso.intent.Intents
import androidx.test.espresso.intent.matcher.IntentMatchers
import org.hamcrest.Matchers
import org.junit.Rule
import org.junit.Test
import org.junit.rules.RuleChain
import org.odk.collect.android.R
import org.odk.collect.android.storage.StoragePathProvider
import org.odk.collect.android.support.pages.FormEntryPage
import org.odk.collect.android.support.pages.ProjectSettingsPage
import org.odk.collect.android.support.rules.FormEntryActivityTestRule
import org.odk.collect.android.support.rules.TestRuleChain.chain
import org.odk.collect.androidtest.RecordedIntentsRule
import java.io.File
import java.io.FileOutputStream
import java.util.Random

class FieldListUpdateTest {
    private var rule: FormEntryActivityTestRule = FormEntryActivityTestRule()

    @get:Rule
    var copyFormChain: RuleChain = chain()
        .around(RecordedIntentsRule())
        .around(rule)

    @Test
    fun relevanceChangeAtEnd_ShouldToggleLastWidgetVisibility() {
        rule.setUpProjectAndCopyForm("fieldlist-updates.xml")
            .fillNewForm("fieldlist-updates.xml", "fieldlist-updates")
            .assertNoQuestion("Target1")
            .answerQuestion("Source1", "A")
            .assertQuestion("Target1")
            .assertTextBelow("Target1", "Source1")
            .longPressOnQuestion("Source1")
            .removeResponse()
            .assertNoQuestion("Target1")
    }

    @Test
    fun relevanceChangeAtBeginning_ShouldToggleFirstWidgetVisibility() {
        rule.setUpProjectAndCopyForm("fieldlist-updates.xml")
            .fillNewForm("fieldlist-updates.xml", "fieldlist-updates")
            .clickGoToArrow()
            .clickGoUpIcon()
            .clickOnGroup("Single relevance at beginning")
            .clickOnQuestion("Source2")
            .assertNoQuestion("Target2")
            .answerQuestion("Source2", "A")
            .assertQuestion("Target2")
            .assertTextBelow("Source2", "Target2")
            .longPressOnQuestion("Source2")
            .removeResponse()
            .assertNoQuestion("Target2")
    }

    @Test
    fun relevanceChangeInMiddle_ShouldToggleMiddleWidgetVisibility() {
        rule.setUpProjectAndCopyForm("fieldlist-updates.xml")
            .fillNewForm("fieldlist-updates.xml", "fieldlist-updates")
            .clickGoToArrow()
            .clickGoUpIcon()
            .clickOnGroup("Single relevance in middle")
            .clickOnQuestion("Source3")
            .assertNoQuestion("Target3")
            .answerQuestion("Source3", "A")
            .assertQuestion("Target3")
            .assertTextBelow("Filler3", "Source3")
            .assertTextBelow("Filler3", "Target3")
            .longPressOnQuestion("Source3")
            .removeResponse()
            .assertNoQuestion("Target3")
    }

    @Test
    fun changeInValueUsedInLabel_ShouldChangeLabelText() {
        rule.setUpProjectAndCopyForm("fieldlist-updates.xml")
            .fillNewForm("fieldlist-updates.xml", "fieldlist-updates")
            .clickGoToArrow()
            .clickGoUpIcon()
            .clickOnGroup("Label change")
            .clickOnQuestion("Hello, , how are you today?")
            .assertQuestion("Hello, , how are you today?")
            .answerQuestion("What is your name?", "Adam")
            .assertQuestion("Hello, Adam, how are you today?")
            .longPressOnQuestion("What is your name?")
            .removeResponse()
            .assertQuestion("Hello, , how are you today?")
    }

    @Test
    fun changeInValueUsedInHint_ShouldChangeHintText() {
        rule.setUpProjectAndCopyForm("fieldlist-updates.xml")
            .fillNewForm("fieldlist-updates.xml", "fieldlist-updates")
            .clickGoToArrow()
            .clickGoUpIcon()
            .clickOnGroup("Hint change")
            .clickOnQuestion("What is your name?")
            .assertText("Please don't use your calculator, !")
            .answerQuestion("What is your name?", "Adam")
            .assertText("Please don't use your calculator, Adam!")
            .longPressOnQuestion("What is your name?")
            .removeResponse()
            .assertText("Please don't use your calculator, !")
    }

    @Test
    fun changeInValueUsedInOtherField_ShouldChangeValue() {
        rule.setUpProjectAndCopyForm("fieldlist-updates.xml")
            .fillNewForm("fieldlist-updates.xml", "fieldlist-updates")
            .clickGoToArrow()
            .clickGoUpIcon()
            .clickOnGroup("Value change")
            .clickOnQuestion("What is your name?")
            .assertAnswer("Name length", "0")
            .assertAnswer("First name letter", "")
            .answerQuestion("What is your name?", "Adam")
            .assertAnswer("Name length", "4")
            .assertAnswer("First name letter", "A")
            .longPressOnQuestion("What is your name?")
            .removeResponse()
            .assertAnswer("Name length", "0")
            .assertAnswer("First name letter", "")
    }

    @Test
    fun selectionChangeAtFirstCascadeLevel_ShouldUpdateNextLevels() {
        rule.setUpProjectAndCopyForm("fieldlist-updates.xml", listOf("fruits.csv"))
            .fillNewForm("fieldlist-updates.xml", "fieldlist-updates")
            .clickGoToArrow()
            .clickGoUpIcon()
            .clickOnGroup("Cascading select")
            .clickOnQuestion("Level1") // No choices should be shown for levels 2 and 3 when no selection is made for level 1
            .assertTextsDoNotExist("A1", "B1", "C1", "A1A") // Selecting C for level 1 should only reveal options for C at level 2
            .clickOnText("C")
            .assertTextsDoNotExist("A1", "B1", "A1A")
            .assertText("C1") // Selecting A for level 1 should reveal options for A at level 2
            .clickOnText("A")
            .assertTextsDoNotExist("A1A", "B1", "C1")
            .assertText("A1") // Selecting A1 for level 2 should reveal options for A1 at level 3
            .clickOnText("A1")
            .assertText("A1A")
            .assertTextsDoNotExist("B1", "C1")
            .longPressOnQuestion("Level1")
            .removeResponse()
            .assertTextsDoNotExist("A1", "B1", "C1", "A1A")
    }

    @Test
    fun clearingParentSelect_ShouldUpdateAllDependentLevels() {
        rule.setUpProjectAndCopyForm("fieldlist-updates.xml", listOf("fruits.csv"))
            .fillNewForm("fieldlist-updates.xml", "fieldlist-updates")
            .clickGoToArrow()
            .clickGoUpIcon()
            .clickOnGroup("Cascading select")
            .clickOnQuestion("Level1")
            .clickOnText("A")
            .clickOnText("A1")
            .clickOnText("A1B")
            .longPressOnQuestion("Level1")
            .removeResponse()
            .assertTextsDoNotExist("A1", "A1B")
    }

    @Test
    fun selectionChangeAtOneCascadeLevelWithMinimalAppearance_ShouldUpdateNextLevels() {
        rule.setUpProjectAndCopyForm("fieldlist-updates.xml", listOf("fruits.csv"))
            .fillNewForm("fieldlist-updates.xml", "fieldlist-updates")
            .clickGoToArrow()
            .clickGoUpIcon()
            .clickOnGroup("Cascading select minimal")
            .clickOnQuestion("Level1")
            .assertTextsDoNotExist("A1", "B1", "C1", "A1A") // No choices should be shown for levels 2 and 3 when no selection is made for level 1
            .openSelectMinimalDialog(0)
            .selectItem("C") // Selecting C for level 1 should only reveal options for C at level 2
            .assertTextsDoNotExist("A1", "B1")
            .openSelectMinimalDialog(1)
            .selectItem("C1")
            .assertTextDoesNotExist("A1A")
            .clickOnText("C")
            .clickOnText("A") // Selecting A for level 1 should reveal options for A at level 2
            .openSelectMinimalDialog(1)
            .assertText("A1")
            .assertTextsDoNotExist("A1A", "B1", "C1")
            .selectItem("A1") // Selecting A1 for level 2 should reveal options for A1 at level 3
            .openSelectMinimalDialog(2)
            .assertText("A1A")
            .assertTextsDoNotExist("B1A", "B1", "C1")
    }

    @Test
    fun questionsAppearingBeforeCurrentTextQuestion_ShouldNotChangeFocus() {
        rule.setUpProjectAndCopyForm("fieldlist-updates.xml")
            .fillNewForm("fieldlist-updates.xml", "fieldlist-updates")
            .clickGoToArrow()
            .clickGoUpIcon()
            .clickOnGroup("Push off screen")
            .clickOnQuestion("Source9")
            .assertNoQuestion("Target9-15")
            .answerQuestion("Source9", "A")
            .assertQuestion("Target9-15")
            .assertQuestionHasFocus("Source9")
    }

    @Test
    fun questionsAppearingBeforeCurrentBinaryQuestion_ShouldNotChangeFocus() {
        rule.setUpProjectAndCopyForm("fieldlist-updates.xml")
            .fillNewForm("fieldlist-updates.xml", "fieldlist-updates")
            .apply {
                // FormFillingActivity expects an image at a fixed path so copy the app logo there
                val icon = BitmapFactory.decodeResource(
                    ApplicationProvider.getApplicationContext<Context>().resources,
                    R.drawable.notes
                )
                val tmpJpg = File(StoragePathProvider().getTmpImageFilePath())
                tmpJpg.createNewFile()
                val fos = FileOutputStream(tmpJpg)
                icon.compress(Bitmap.CompressFormat.JPEG, 90, fos)
                Intents.intending(Matchers.not(IntentMatchers.isInternal())).respondWith(
                    Instrumentation.ActivityResult(
                        Activity.RESULT_OK, null
                    )
                )
            }
            .clickGoToArrow()
            .clickGoUpIcon()
            .clickOnGroup("Push off screen binary")
            .clickOnQuestion("Source10")
            .assertNoQuestion("Target10-15")
            .clickOnString(org.odk.collect.strings.R.string.capture_image)
            .assertQuestion("Target10-15")
            .assertText(org.odk.collect.strings.R.string.capture_image)
    }

    @Test
    fun changeInValueUsedInGuidanceHint_ShouldChangeGuidanceHintText() {
        rule.setUpProjectAndCopyForm("fieldlist-updates.xml")
            .fillNewForm("fieldlist-updates.xml", "fieldlist-updates")
            .clickOptionsIcon()
            .clickProjectSettings()
            .clickFormManagement()
            .openShowGuidanceForQuestions()
            .clickOnString(org.odk.collect.strings.R.string.guidance_yes)
            .pressBack(ProjectSettingsPage())
            .pressBack(FormEntryPage("fieldlist-updates"))
            .clickGoToArrow()
            .clickGoUpIcon()
            .clickOnGroup("Guidance hint")
            .clickOnQuestion("Source11")
            .assertTextDoesNotExist("10")
            .answerQuestion("Source11", "5")
            .assertQuestion("10")
            .longPressOnQuestion("Source11")
            .removeResponse()
            .assertTextDoesNotExist("10")
    }

    @Test
    fun selectingADateForDateTime_ShouldChangeRelevanceOfRelatedField() {
        rule.setUpProjectAndCopyForm("fieldlist-updates.xml")
            .fillNewForm("fieldlist-updates.xml", "fieldlist-updates")
            .clickGoToArrow()
            .clickGoUpIcon()
            .clickOnGroup("Date time")
            .clickOnQuestion("Source12")
            .assertNoQuestion("Target12")
            .clickOnString(org.odk.collect.strings.R.string.select_date)
            .clickOKOnDialog()
            .assertQuestion("Target12")
            .longPressOnQuestion("Source12")
            .removeResponse()
            .assertNoQuestion("Target12")
    }

    @Test
    fun selectingARating_ShouldChangeRelevanceOfRelatedField() {
        rule.setUpProjectAndCopyForm("fieldlist-updates.xml")
            .fillNewForm("fieldlist-updates.xml", "fieldlist-updates")
            .clickGoToArrow()
            .clickGoUpIcon()
            .clickOnGroup("Rating")
            .clickOnQuestion("Source13")
            .assertNoQuestion("Target13")
            .setRating(3.0f)
            .assertQuestion("Target13")
            .longPressOnQuestion("Source13")
            .removeResponse()
            .assertNoQuestion("Target13")
    }

    @Test
    fun manuallySelectingAValueForMissingExternalApp_ShouldTriggerUpdate() {
        rule.setUpProjectAndCopyForm("fieldlist-updates.xml")
            .fillNewForm("fieldlist-updates.xml", "fieldlist-updates")
            .clickGoToArrow()
            .clickGoUpIcon()
            .clickOnGroup("External app")
            .clickOnQuestion("Source14")
            .clickOnText("Launch")
            .assertNoQuestion("Target14")
            .answerQuestion("Source14", Random().nextInt().toString())
            .assertQuestion("Target14")
            .longPressOnQuestion("Source14")
            .removeResponse()
            .assertNoQuestion("Target14")
    }

    @Test
    fun searchMinimalInFieldList() {
        rule.setUpProjectAndCopyForm("fieldlist-updates.xml", listOf("fruits.csv"))
            .fillNewForm("fieldlist-updates.xml", "fieldlist-updates")
            .clickGoToArrow()
            .clickGoUpIcon()
            .clickOnGroup("Search in field-list")
            .clickOnQuestion("Source15")
            .assertSelectMinimalDialogAnswer(null)
            .assertNoQuestion("Target15")
            .openSelectMinimalDialog()
            .assertTexts("Mango", "Oranges", "Strawberries")
            .selectItem("Strawberries")
            .assertQuestion("Target15")
            .assertSelectMinimalDialogAnswer("Strawberries")
            .longPressOnQuestion("Source15")
            .removeResponse()
            .assertSelectMinimalDialogAnswer(null)
            .assertNoQuestion("Target15")
    }

    @Test
    fun listOfQuestionsShouldNotBeScrolledToTheLastEditedQuestionAfterClickingOnAQuestion() {
        rule.setUpProjectAndCopyForm("fieldlist-updates.xml")
            .fillNewForm("fieldlist-updates.xml", "fieldlist-updates")
            .clickGoToArrow()
            .clickGoUpIcon()
            .clickOnGroup("Long list of questions")
            .clickOnQuestion("Question1")
            .answerQuestion(0, "X")
            .clickOnQuestionField("Question20")
            .assertQuestion("Question20")
    }

    @Test
    fun recordingAudio_ShouldChangeRelevanceOfRelatedField() {
        rule.setUpProjectAndCopyForm("fieldlist-updates.xml")
            .fillNewForm("fieldlist-updates.xml", "fieldlist-updates")
            .clickGoToArrow()
            .clickGoUpIcon()
            .clickOnGroup("Audio")
            .clickOnQuestion("Source16")
            .assertNoQuestion("Target16")
            .clickOnString(org.odk.collect.strings.R.string.capture_audio)
            .clickOnContentDescription(org.odk.collect.strings.R.string.stop_recording)
            .assertQuestion("Target16")
            .clickOnString(org.odk.collect.strings.R.string.delete_answer_file)
            .clickOnTextInDialog(
                org.odk.collect.strings.R.string.delete_answer_file,
                FormEntryPage("fieldlist-updates")
            )
            .assertNoQuestion("Target16")
    }

    @Test
    fun changeInValueUsedToDetermineIfAQuestionIsRequired_ShouldUpdateTheRelatedRequiredQuestion() {
        rule.setUpProjectAndCopyForm("dynamic_required_question.xml")
            .fillNewForm("dynamic_required_question.xml", "dynamic_required_question")
            .assertQuestion("Target", false)
            .answerQuestion("Source", "blah")
            .assertQuestion("Target", true)
            .swipeToNextQuestionWithConstraintViolation(org.odk.collect.strings.R.string.required_answer_error)
            .longPressOnQuestion("Source")
            .removeResponse()
            .assertQuestion("Target", false)
            .assertTextDoesNotExist(org.odk.collect.strings.R.string.required_answer_error)
    }
}
